Sources/Container.swift

```diff
@@ -293,32 +308,42 @@ extension Container: Resolver {
 
     fileprivate func resolve<Service, Factory>(
         entry: ServiceEntryProtocol,
-        invoker: (Factory) -> Any
+        invoker: @escaping (Factory) -> Any
     ) -> Service? {
-        incrementResolutionDepth()
-        defer { decrementResolutionDepth() }
-
-        guard let currentObjectGraph = currentObjectGraph else {
-            fatalError("If accessing container from multiple threads, make sure to use a synchronized resolver.")
-        }
-
-        if let persistedInstance = persistedInstance(Service.self, from: entry, in: currentObjectGraph) {
-            return persistedInstance
-        }
-
-        let resolvedInstance = invoker(entry.factory as! Factory)
-        if let persistedInstance = persistedInstance(Service.self, from: entry, in: currentObjectGraph) {
-            // An instance for the key might be added by the factory invocation.
-            return persistedInstance
+        let resolution: () -> Service? = { [weak self] in
```

Comment:
Maybe strong [self] is okay to capture here? Reference will be release when we exit scope and that way can omit all the `self.*` for readability.

Reply:
I have no objections to that - makes sense since it's within the same scope. Will wait for further comments before I make that change

Reply:
Actually on looking at this again, I think it should stay `weak`.

In the event that the thread is locked waiting for the lock to be acquired and `self` is deallocated, this should exit without returning a value IMO.

Reply:
Is it possible that the compiler will allow for `Container` to be deinitialized before `resolve<Service, Factory>` exits? I ask because all class-level functions are passed `self` as an argument at a lower level.

Reply:
Hmm... you might be right. Because we're locking on the execution of the function, the reference to self is likely preserved either way. Updated

Reply:
Ok—could you remove the `self.` references from within the block for style? It's fine if you keep them just figured they're redundant now